src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -178,24 +171,31 @@ protected override async Task OnRun(CancellationToken token)
                         }
                         finally
                         {
+                            if (_stateHolder.ActionExecutionSucceeded(actionsCompleted))
+                            {
+                                Context.Logger.CollectionRuleActionsCompleted(Context.Name);
+                            }
+
+                            _stateHolder.CanExecuteActions(Context.Clock.UtcNow.UtcDateTime);
+
                             // The collection rule has executed the action list the maximum
                             // number of times as specified by the limits and the action count
                             // window was not specified. Since the pipeline can no longer execute
                             // actions, the pipeline can complete.
-                            completePipeline = actionCountLimit <= executionTimestamps.Count &&
-                                !actionCountWindowDuration.HasValue;
-                        }
+                            completePipeline = _stateHolder.ActionCountLimit <= _stateHolder.ExecutionTimestamps.Count &&
```

Comment:
Consider pushing logic that operates against _stateHolder fields to Instance methods on StateHolder that you can then lock against.

Reply:
Since we're not operating on `ExecutionTimestamps` (and just getting its count), I don't believe we'll need to lock - but I can move this logic into `CollectionRulePipelineState`, which will let us make `ActionCountReached()` a private method.